### PR TITLE
Fix the Anilist .get() regression

### PR DESCRIPTION
**Pull request by Nihilate**
See #55 for more details, but the tl;dr is:

`Traceback (most recent call last):
  File "Z:\xxxx\Roboragi\roboragi\Anilist.py", line 361, in getClosestManga
    title_english = manga.get('title_english', '').lower() or None
AttributeError: 'NoneType' object has no attribute 'lower'`

"title_english" exists, but is None, causing the program to fail when lowering the result.

### CODE REVIEW

**Comment by Nihilate** [comment]
Can you think of a nicer way of doing this @reillysiemens? The whole thing is a little messy and will eventually be removed in #24, so I don't want to worry about it _too_ much.

File: roboragi/Anilist.py [hunk 1]
```diff
@@ -291,8 +291,8 @@ def getClosestAnime(searchText, animeList):
         if closestNameFromList:
             closestNameFromList = closestNameFromList.lower()
             for anime in animeList:
-                title_english = anime.get('title_english', '').lower()
-                title_romaji = anime.get('title_english', '').lower()
+                title_english = (anime.get('title_english') or '').lower()
+                title_romaji = (anime.get('title_romaji') or '').lower()
```
Comment (reillysiemens) [inline roboragi/Anilist.py]: Oof. I did not mean to break that. :sweat_smile: Copy + paste is dangerous!

File: roboragi/Anilist.py [hunk 2]
```diff
@@ -356,10 +356,12 @@ def getClosestManga(searchText, mangaList, isLN=False):
                 mangaList.remove(manga)
 
         for manga in mangaList:
-            title_english = manga.get('title_english', '').lower() or None
-            title_romaji = manga.get('title_romaji', '').lower() or None
-            mangaNameList.append(title_english)
-            mangaNameList.append(title_romaji)
+            title_english = manga.get('title_english')
+            title_romaji = manga.get('title_romaji')
+            mangaNameList.append(title_english.lower())\
```
Comment (reillysiemens) [inline roboragi/Anilist.py]: If this line got expanded out it'd be
```python
mangaNameList.append(title_english.lower()) if title_english else None
```
which, while valid Python (you aren't required to use values from function calls or expressions), is not pretty.

I think, in this case, the best way to handle this is to just use a plain ol' conditional.
```python
if title_english:
    mangaNameList.append(title_english.lower())
```

File: roboragi/Anilist.py [hunk 3]
```diff
@@ -379,8 +381,8 @@ def getClosestManga(searchText, mangaList, isLN=False):
             closestNameFromList = closestNameFromList.lower()
             for manga in mangaList:
                 if not ('one shot' in manga['type'].lower()):
-                    title_english = manga.get('title_english', '').lower()
-                    title_romaji = manga.get('title_english', '').lower()
+                    title_english = (manga.get('title_english') or '').lower()
+                    title_romaji = (manga.get('title_romaji') or '').lower()
```
Comment (reillysiemens) [inline roboragi/Anilist.py]: Yup. Did it here too. :upside_down_face: Sorry about that...

File: roboragi/Anilist.py (hunk 2 above)
Comment (Nihilate) [reply roboragi/Anilist.py]: Sounds good to me, I'll make the change when I get home.

**Comment by Nihilate** [comment]
Yeah, there are definitely a few gotchas with some of the returned values. I'm sure there's a bunch of weird looking code in there which is weird for reasons long forgotten because of it.

If you wanted to run some "live" sanity tests locally, the only data source you'll need a username and password for will be VNDB. I'd like to get some integration tests working eventually, but they should probably be kept away from the unit tests since you don't want a build failing if a website is down. Gut feeling is to set them up to be run in a pipeline to run once a day, with an issue/email on failure.

**Comment by reillysiemens** [comment]
Yeah, unit tests and integration tests are both important. I agree it'd be better to have integration tests in this scenario run as part of a daily pipeline.

I've been thinking about this regression over the last few hours and the conclusion that I've come to is that without unit tests at least we'll probably run into more issues like this as we try to make even small fixes like increasing PEP8 compliance.

I'll see what I can do to get some manual sanity tests going. As expected, verification will probably slow the rate of contributions, but better that than breaking production. And maybe I'll be able to build some tests as I go once I know more about the expected output.

**Comment by Nihilate** [comment]
Sounds good. I might focus on getting a basic test suite of integration tests up for each data source after the config changes/containerisation, since I've got the most context around what to expect, but feel free to add some if you get there before me. They'll be worth having for regression alone, but it'll be good to have a baseline to work off while the data sources get refactored.

**Pull request merged**